### PR TITLE
Incomplete Cholesky factorizations and preconditioners

### PR DESCRIPTION
**Pull request by upsj**
This PR adds Incomplete Cholesky preconditioners (both parallel asynchronous as well as exact) to Ginkgo

### CODE REVIEW

**Comment by upsj** [comment]
check-format!

**Comment by upsj** [comment]
check-format!

**Comment by upsj** [comment]
check-format!

**Comment by upsj** [comment]
We seem to hit some size limits with MinGW/Cygwin Debug builds, so I will have to increase the inlining level there.

**Review by yhmtsai — COMMENTED** [review]
first part of my review

File: reference/factorization/par_ic_kernels.cpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    auto l_col_idxs = l->get_const_col_idxs();
+    auto l_vals = l->get_values();
+    auto a_vals = a_lower->get_const_values();
+
+    for (size_type row = 0; row < num_rows; ++row) {
+        for (size_type l_nz = l_row_ptrs[row]; l_nz < l_row_ptrs[row + 1];
+             ++l_nz) {
+            auto col = l_col_idxs[l_nz];
+            auto a_val = a_vals[l_nz];
+            // accumulate l(row,:) * l(col,:) without the last entry l(col, col)
```
Comment (yhmtsai) [inline reference/factorization/par_ic_kernels.cpp:100]: Is there any reason we do the last entry computation and then remove it in ParILU?
Or, is it just different implementation style?

**Review by yhmtsai — APPROVED** [review]
LGTM

**Review by Slaedr — COMMENTED** [review]
Excellent work! I think it's almost ready, I just have some small comments and requests.

**Review by pratikvn — COMMENTED** [review]
First part of review.

File: reference/test/factorization/par_ic_kernels.cpp [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+#include <memory>
+#include <vector>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
```
Comment (tcojean) [inline reference/test/factorization/par_ic_kernels.cpp:46]:
```suggestion
#include <ginkgo/core/matrix/csr.hpp>
#include <ginkgo/core/matrix/dense.hpp>
```

**Review by tcojean — APPROVED** [review]
LGTM. Some minor comments.

File: benchmark/utils/preconditioners.hpp [hunk 3]
```diff
@@ -48,14 +48,14 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+              "Supported values are: none, jacobi, paric, parict, parilu, "
+              "parilut, ic, ilu, paric-isai, parict-isai, parilu-isai, "
+              "parilut-isai, ic-isai, ilu-isai, overhead");
 
 DEFINE_uint32(parilu_iterations, 5,
-              "The number of iterations for ParICT/ParILU(T)");
+              "The number of iterations for ParIC(T)/ParILU(T)");
 
 DEFINE_bool(parilut_approx_select, true,
             "Use approximate selection for ParICT/ParILUT");
```
[the remark this answers is not in the document]
Comment (upsj) [reply benchmark/utils/preconditioners.hpp:61]: Yes, this parameter is only used for threshold-selection in ParILUT and ParICT

File: cuda/test/factorization/ic_kernels.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    std::shared_ptr<gko::ReferenceExecutor> ref;
+    std::shared_ptr<gko::CudaExecutor> cuda;
+    std::ranlux48 rand_engine;
+    std::shared_ptr<Csr> csr_ref;
+    std::shared_ptr<Csr> csr_cuda;
+
+    Ic()
+        : ref(gko::ReferenceExecutor::create()),
+          cuda(gko::CudaExecutor::create(0, ref)),
+          rand_engine(6794)
```
[the remark this answers is not in the document]
Comment (upsj) [reply cuda/test/factorization/ic_kernels.cpp:72]: Just random key smashing, so we have reproducible inputs in the same test, but different inputs in different tests.

File: examples/par-ilu-convergence/doc/results.dox [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+Usage: executable <reference|omp|cuda|hip|dpcpp> [<matrix-file>] [<parilu|parilut|paric|parict] [<max-iterations>] [<num-repetitions>] [fill-in-limit]
+
+@endcode
+
+When specifying an executor:
+
+@code{.cpp}
+
+Reading data/A.mtx
+1;71800;10300;8800;8200;8000;7700;7500;7500;7500;7400;1.0331e-14;1.0331e-14;1.0331e-14;1.0331e-14;1.0331e-14;1.0331e-14;1.0331e-14;1.0331e-14;1.0331e-14;1.0331e-14;
```
[the remark this answers is not in the document]
Comment (upsj) [reply examples/par-ilu-convergence/doc/results.dox:15]: Yes, on the reference executor, the results are exact. I should note though that this is the residual only on the sparsity pattern of L+U, hence the almost zero value.

File: include/ginkgo/core/factorization/par_ic.hpp [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+ * $
+ *
+ * In general, the entries of $L$ can be iterated in parallel and in
+ * asynchronous fashion, the algorithm asymptotically converges to the
+ * incomplete factors $L$ and $L^H$ fulfilling $\left(R = A - L \cdot
+ * L^H\right)\vert_\mathcal{S} = 0\vert_\mathcal{S}$ where $\mathcal{S}$ is the
+ * pre-defined sparsity pattern (in case of IC(0) the sparsity pattern of the
+ * system matrix $A$). The number of ParIC sweeps needed for convergence
+ * depends on the parallelism level: For sequential execution, a single sweep
+ * is sufficient, for fine-grained parallelism, 3 sweeps are typically
```
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/factorization/par_ic.hpp:80]: Probably not, this is mostly copied over from the ParILU documentation.

File: reference/factorization/par_ic_kernels.cpp (hunk 1 above)
Comment (upsj) [reply reference/factorization/par_ic_kernels.cpp:100]: Yes, I added the `l_col < last_entry` condition to remove the last term

File: common/factorization/par_ic_kernels.hpp.inc [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void ic_sweep(
+    const IndexType *__restrict__ a_row_idxs,
+    const IndexType *__restrict__ a_col_idxs,
+    const ValueType *__restrict__ a_vals,
+    const IndexType *__restrict__ l_row_ptrs,
+    const IndexType *__restrict__ l_col_idxs, ValueType *__restrict__ l_vals,
+    IndexType l_nnz)
+{
+    auto l_nz = thread::get_thread_id_flat<IndexType>();
```
[the remark this answers is not in the document]
Comment (upsj) [reply common/factorization/par_ic_kernels.hpp.inc:64]: Good point! `const auto` is more appropriate here, as size_type is unsigned, but IndexType is signed.

File: cuda/base/cusparse_bindings.hpp [hunk 8]
```diff
@@ -1352,6 +1366,104 @@ GKO_BIND_CUSPARSE_ILU0(std::complex<double>, cusparseZcsrilu02);
 #undef GKO_BIND_CUSPARSE_ILU0
 
 
+template <typename ValueType, typename IndexType>
+void ic0_buffer_size(cusparseHandle_t handle, IndexType m, IndexType nnz,
```
[the remark this answers is not in the document]
Comment (upsj) [reply cuda/base/cusparse_bindings.hpp:1370]: To stay consistent with the cuSPARSE interfaces and the other `_buffer_size` functions, I will keep this as-is for now, but this is a good point for the next round of clean-up 😄

File: include/ginkgo/core/preconditioner/ic.hpp [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+    mutable struct cache_struct {
+        cache_struct() = default;
+        ~cache_struct() = default;
+        cache_struct(const cache_struct &) {}
+        cache_struct(cache_struct &&) {}
+        cache_struct &operator=(const cache_struct &) { return *this; }
+        cache_struct &operator=(cache_struct &&) { return *this; }
+        std::unique_ptr<LinOp> intermediate{};
+    } cache_;
+};  // namespace preconditioner
```
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/preconditioner/ic.hpp:389]: exactly 😆 I had a missing closing brace when saving the file a while back, which added the comment, but didn't remove it after being fixed.

File: reference/test/preconditioner/ic.cpp [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+    auto before_lh_solver = ic->get_lh_solver();
+
+    auto clone = ic->clone();
+
+    ASSERT_EQ(before_l_solver, clone->get_l_solver());
+    ASSERT_EQ(before_lh_solver, clone->get_lh_solver());
+}
+
+
+TYPED_TEST(Ic, CanBeTransposed)
```
[the remark this answers is not in the document]
Comment (upsj) [reply reference/test/preconditioner/ic.cpp:230]: True, we are missing a large number of tests for complex numbers altogether. But for now, I think we can keep this in place?

File: reference/test/preconditioner/ic.cpp [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+    auto before_l_solver = ic->get_l_solver();
+    auto before_lh_solver = ic->get_lh_solver();
+    // The switch up of matrices is intentional, to make sure they are distinct!
+    auto lh_l_composition =
+        Composition::create(this->lh_factor, this->l_factor);
+    auto moved = ic_prec_type::build()
+                     .on(this->exec)
+                     ->generate(gko::share(lh_l_composition));
+
+    moved->copy_from(std::move(ic));
```
[the remark this answers is not in the document]
Comment (upsj) [reply reference/test/preconditioner/ic.cpp:210]: ic->move_to might be appropriate, which afaik does the same thing, but only for a known static type. I think I had to write is this way due to a mismatch between dynamic and static type.

File: reference/factorization/par_ic_kernels.cpp [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+            auto lh_begin = l_row_ptrs[col];
+            auto lh_end = l_row_ptrs[col + 1];
+            while (l_begin < l_end && lh_begin < lh_end) {
+                auto l_col = l_col_idxs[l_begin];
+                auto lh_row = l_col_idxs[lh_begin];
+                if (l_col == lh_row && l_col < col) {
+                    sum += l_vals[l_begin] * conj(l_vals[lh_begin]);
+                }
+                if (lh_row == row) {
+                    lh_nz = lh_begin;
```
[the remark this answers is not in the document]
Comment (upsj) [reply reference/factorization/par_ic_kernels.cpp:114]: Good catch, this is left-over from the ParICT kernel implementation I based this kernel on.

File: reference/factorization/ic_kernels.cpp [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+ * @brief The ic factorization namespace.
+ *
+ * @ingroup factor
+ */
+namespace ic_factorization {
+
+
+template <typename ValueType, typename IndexType>
+void compute(std::shared_ptr<const DefaultExecutor> exec,
+             matrix::Csr<ValueType, IndexType> *m) GKO_NOT_IMPLEMENTED;
```
[the remark this answers is not in the document]
Comment (upsj) [reply reference/factorization/ic_kernels.cpp:49]: Yes, this mirrors the `ilu` factorization which is only available on GPUs based on cu/hipSPARSE

File: include/ginkgo/core/preconditioner/ic.hpp [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+
+        return std::move(transposed);
+    }
+
+protected:
+    void apply_impl(const LinOp *b, LinOp *x) const override
+    {
+        set_cache_to(b);
+        l_solver_->apply(b, cache_.intermediate.get());
+        x->copy_from(cache_.intermediate.get());
```
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/preconditioner/ic.hpp:199]: Good point, I added a check.

File: cuda/test/factorization/ic_kernels.cpp [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+                        .with_skip_sorting(true)
+                        .on(ref)
+                        ->generate(csr_ref);
+    auto cuda_fact = gko::factorization::Ic<>::build()
+                         .with_skip_sorting(true)
+                         .on(cuda)
+                         ->generate(csr_cuda);
+
+    GKO_ASSERT_MTX_NEAR(ref_fact->get_l_factor(), cuda_fact->get_l_factor(),
+                        1e-14);
```
[the remark this answers is not in the document]
Comment (upsj) [reply cuda/test/factorization/ic_kernels.cpp:102]: We already have a struct for this: `r<ValueType>::value`. I don't use it yet, but I have a few test improvements in mind that this kind of change would nicely fit into :)

File: include/ginkgo/core/preconditioner/ic.hpp [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+
+        // If no factories are provided, generate default ones
+        if (!parameters_.l_solver_factory) {
+            l_solver_ = generate_default_solver<l_solver_type>(exec, l_factor);
+            // If comp contains both factors: use the transposed factor to avoid
+            // transposing twice
+            if (comp->get_operators().size() == 2) {
+                auto lh_factor = comp->get_operators()[1];
+                GKO_ASSERT_EQUAL_DIMENSIONS(l_factor, lh_factor);
+                lh_solver_ = as<lh_solver_type>(l_solver_->conj_transpose());
```
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/preconditioner/ic.hpp:258]: Yes, the reason for this is that we don't have a uniform way to specify "this factory commutes with transposition", which is true for solvers, but not for ISAI, so I left it like this for now.

File: examples/par-ilu-convergence/par-ilu-convergence.cpp [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+    } else if (precond == "paric") {
+        factory = gko::factorization::ParIc<>::build().on(exec);
+        set_iterations = [&](int it) {
+            gko::as<gko::factorization::ParIc<>::Factory>(factory)
+                ->get_parameters()
+                .iterations = it;
+        };
+    } else if (precond == "parilut") {
+        factory =
+            gko::factorization::ParIlut<>::build().with_fill_in_limit(limit).on(
```
[the remark this answers is not in the document]
Comment (upsj) [reply examples/par-ilu-convergence/par-ilu-convergence.cpp:148]: This is related to the discussions in #472, I would like to address these issues with 2.0

File: include/ginkgo/core/preconditioner/ic.hpp [hunk 18]
```diff
@@ -0,0 +1,396 @@
...
+     * @internal  The second template parameter (which uses SFINAE) must
+     * match the default value of the general case in order to be accepted
+     *            as a specialization, which is why `xstd::void_t` is used.
+     */
+    template <typename SolverType>
+    struct has_with_criteria<
+        SolverType,
+        xstd::void_t<decltype(std::declval<factory_type_t<SolverType>>()
+                                  .with_criteria(with_criteria_param_type()))>>
+        : std::true_type {};
```
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/preconditioner/ic.hpp:324]: Yes, I moved them to "solver/solver_traits.hpp", how does that sound?

File: include/ginkgo/core/preconditioner/ic.hpp [hunk 19]
```diff
@@ -0,0 +1,396 @@
...
+     * can, it is assigned default values which should be well suited for a
+     * preconditioner.
+     */
+    template <typename SolverType>
+    static std::enable_if_t<has_with_criteria<SolverType>::value,
+                            std::unique_ptr<SolverType>>
+    generate_default_solver(const std::shared_ptr<const Executor> &exec,
+                            const std::shared_ptr<const LinOp> &mtx)
+    {
+        constexpr gko::remove_complex<value_type> default_reduce_residual{1e-4};
```
[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/preconditioner/ic.hpp:340]: I don't have much experience using iterative solvers for triangular solves, maybe @fritzgoebel @hartwiganzt can chime in here?

File: cuda/test/factorization/ic_kernels.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (Slaedr) [reply cuda/test/factorization/ic_kernels.cpp:102]: Alright. Side note: I don't like these one-letter identifiers. 😅

File: reference/test/preconditioner/ic.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (Slaedr) [reply reference/test/preconditioner/ic.cpp:230]: Yeah, I guess so.

File: include/ginkgo/core/preconditioner/ic.hpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (Slaedr) [reply include/ginkgo/core/preconditioner/ic.hpp:256]: Ah, you want to guarantee that the IC is always a symmetric/hermitian preconditioner?

[the remark this answers is not in the document]
Comment (upsj) [reply include/ginkgo/core/preconditioner/ic.hpp:256]: Yes, we want an incomplete factorization preconditioner that is safe to use together with spd solvers independent of its internal solver (exact, ISAI, iterative, ...)

**Pull request merged**